libavpipe/src/avpipe_xc.c

```diff
@@ -383,25 +425,38 @@ prepare_decoder(
     const AVIOInterruptCB int_cb = { decode_interrupt_cb, (void*)decoder_context};
     decoder_context->format_context->interrupt_callback = int_cb;
 
+    decoder_context->live_proto = find_live_proto(inctx);
+
     /* Set our custom reader */
     prepare_input(in_handlers, inctx, decoder_context, seekable);
 
     AVDictionary *opts = NULL;
-    if (params && params->listen && (decoder_context->is_rtmp || decoder_context->is_srt))
+    if (params && params->listen && is_live_source(decoder_context))
         av_dict_set(&opts, "listen", "1" , 0);
 
-    if (decoder_context->is_rtmp &&
-        params->listen &&
-        params->connection_timeout > 0) {
-        char timeout[32];
-        sprintf(timeout, "%d", params->connection_timeout);
-        av_dict_set(&opts, "timeout", timeout, 0);
-    } else if (decoder_context->is_srt && params->listen && params->connection_timeout > 0) {
+    if (params->listen && params->connection_timeout > 0) {
         char timeout[32];
         int64_t connection_timeout_micros = MICRO_IN_SEC * (int64_t)params->connection_timeout;
-        sprintf(timeout, "%"PRId64, connection_timeout_micros);
-        /* SRT timeout is in microseconds */
-        av_dict_set(&opts, "listen_timeout", timeout, 0);
+
+        switch(decoder_context->live_proto) {
+        case avp_proto_rtmp:
+            sprintf(timeout, "%d", params->connection_timeout);
+            av_dict_set(&opts, "timeout", timeout, 0);
+            break;
+        case avp_proto_srt:
+            /* SRT timeout is in microseconds */
+            sprintf(timeout, "%"PRId64, connection_timeout_micros);
+            av_dict_set(&opts, "listen_timeout", timeout, 0);
+            break;
+        case avp_proto_rtp:
+            // RTP timeout is in microseconds
```

Comment:
We should note in the comment here that this means that the stream _will not time out_ even after initially connected until after this timeout. I wish these parameters were handled separately in ffmpeg.

Reply:
I actually looked into how hard that would be to add, and it was pretty easy, so here's a PR that implements a separate `listen_timeout` in ffmpeg's RTP protocol: https://github.com/eluv-io/FFmpeg/pull/12

I recommend reviewing it by looking at exactly how the normal `rw_timeout` is used, as then you can see what's going on.